packages/core/src/currentScopes.ts

```diff
@@ -45,3 +49,76 @@ export function getGlobalScope(): Scope {
 export function setGlobalScope(scope: Scope | undefined): void {
   globalScope = scope;
 }
+
+/**
+ * Creates a new scope with and executes the given operation within.
+ * The scope is automatically removed once the operation
+ * finishes or throws.
+ */
+export function withScope<T>(callback: (scope: Scope) => T): T;
+/**
+ * Set the given scope as the active scope in the callback.
+ */
+export function withScope<T>(scope: Scope | undefined, callback: (scope: Scope) => T): T;
+/**
+ * Either creates a new active scope, or sets the given scope as active scope in the given callback.
+ */
+export function withScope<T>(
+  ...rest: [callback: (scope: Scope) => T] | [scope: Scope | undefined, callback: (scope: Scope) => T]
+): T {
```

Comment:
I know UX-wise passing the scope in first if defined feels better, but IMO I think this will lead to bugs/confusing code. I'd prefer if we just always had scope as the second argument.

```suggestion
export function withScope<T>(
  callback: callback: (scope: Scope) => T,
  scope: Scope | undefined,
): T {
```

Reply:
This is a bit tricky because we already have this API in v7 etc. So I'd leave this for now, we can still think about adjusting this but then we need to backport this to v7, deprecate the deprecation again, etc 😬

Reply:
Note that in the ACS implementation, these are actually separate by design, so we can (if we want to) eventually change this user-facing without having to touch the ACS stuff.